src/csharp/Grpc.Core/ChannelCredentials.cs

```diff
@@ -104,40 +106,67 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
         }
     }
 
+    /// <summary>
+    /// Callback invoked with the expected targetHost and the peer's certificate.
+    /// If false is returned by this callback then it is treated as a
+    /// verification failure. Invocation of the callback is blocking, so any
+    /// implementation should be light-weight.
+    /// </summary>
+    /// <param name="context">The <see cref="T:Grpc.Core.VerifyPeerContext"/> associated with the callback</param>
+    /// <returns>true if verification succeeded, false otherwise.</returns>
+    /// Note: experimental API that can change or be removed without any prior notice.
```

Comment:
We should probably document the thread safety of this API.... looking at the invocation of the callback from C-core, it seems it's not thread safe, which might be counter-intuitive to users of this API (this could be a real concern e.g. in the case when we're connecting to a service in round robin and this callback is invoked multiple times all at once).

Might also be worth considering making it thread safe at the C# layer.

Reply:
what exactly do you mean by the API not being thread safe?

Reply:
I just mean that, authors of the user-provided `VerifyPeerCallback` delegate need to be prepared for their callback to potentially be invoked multiple times, concurrently from different threads.

Reply:
at least, from looking <i>I think</i> they need to be prepared for that

Reply:
Good point. Added that to the doc comment.